app/components/web-components/vl-datepicker/component.js

```diff
@@ -21,7 +21,11 @@ export default Component.extend({
   selectedDate: computed('date', function () {
     const date = this.get('date');
     if (date) {
-      return this.formatter.formatDate(date.get('firstObject'));
+      if (date instanceof Date) {
+        return date;
+      } else {
```

Comment:
Oh boy, I can't say I like this bit. We're bridging a gap here that shouldn't have been. The date here - which is eventually passed down to formatters - is expected to be a `formatted date string`. I'd keep it like that. What went wrong with the original code that made you add this?

Reply:
I was very surprised to find out that our **date**picker component didn't support native Javascript **date** objects (flatpicker [does](https://shipshapecode.github.io/ember-flatpickr/docs/api/components/ember-flatpickr#date)).  
A "formatted date string" is a serialized format. Imo it is a bad practice to serialize stuff as long as you don't really need to (a native object representation is available). I tried to find a reference clearly explaining this, but couldn't find something nice at first glance ... there are many disadvantages to this, but the baseline is that it just makes things more complicated then need be. Passing this formatted date string is exactly the same than if you would pass `JSON.stringify(person)` instead of the `person`-object to an imaginary `person`-component. We don't do that either?
The choice for `Date` vs `Moment` is more subtle. In most of the app, the "interfacing-type" for sharing dates to components, ember-data, ... is a native Javascript Date object. That could be a `moment` two, but from a "implicit documentation" or portability perspective, a `Date` object is the most logical choice in Javascript-land. This doesn't mean you shouldn't use moment ... moment has a whole lot of functionality that the native date object doesn't have, but the information they represent is the same, so for interfacing you'd try to use what "the one on the other end" expects the most ... does all of this make some sense?

Reply:
It certainly makes sense, and in the sense of serialisation, I do get your point. But the benefits of using a generic standard for easier-to-read code overrule the negative side-effects of it imho. This won't slow us down or put us knee-deep in some kind of trouble or something. The native JS Date API also doesn't really play very nice. This is the big reason many developers have grown to really like libraries like moment.js (people like myself). Right now we're using the two together, which makes it hard to understand how code you've never seen works, especially if you're not sure what kind of object you're dealing with. I generally assumed it would be string timestamps until I saw the `|| new Date();`. The original code worked out fine, and this seems to patch up to support the off-chance of native Date objects. It really makes this code feel unnecessarily complex for some reason. I'm all in for readable straightforward code in the first place. I had to go back and forth for this to really sink in. Those are just my thoughts here.

Reply:
Personally I must say that I don't mind the idea of the date data being transferred in a very basic format (string). This has been proven to work in a lot of systems (I guess my cobol background).  It will also keep working when Date or Moment api's change (isn't it an iso standard?).
Even though I have a love hate relationship with moment (undefined returns a valid date), I am a fan of using a library consistently when available. 

If we are going to keep the different formats around: a suggestion:
Maybe it is a good idea to name your date variables so there is no doubt which type it is?  
This might be a bit extreme: https://en.wikipedia.org/wiki/Hungarian_notation but the idea can make it easier to see what kind of variables you are looking at? 
This could give you: nativeDateStartDate or momentStartDate. 
If you don't like this either we should eliminate the need for it.